graylog2-server/src/main/java/org/graylog/mcp/server/Tool.java

```diff
@@ -110,10 +137,30 @@ public Optional<Map<String, Object>> outputSchema() {
      * @param parameterMap raw parameter map
      * @return the return value of the tool call
      */
-    public O apply(PermissionHelper permissionHelper, Map<String, Object> parameterMap) {
+    public ToolResult<O> apply(PermissionHelper permissionHelper, Map<String, Object> parameterMap) {
         final P p = objectMapper.convertValue(parameterMap, parameterType);
-        return apply(permissionHelper, p);
+        if (!isStructuredOutputSet()) {
+            return new ToolResult.Text<>(applyAsText(permissionHelper, p));
+        }
+        return new ToolResult.Data<>(apply(permissionHelper, p));
     }
 
     protected abstract O apply(PermissionHelper permissionHelper, P parameters);
+
+    protected String applyAsText(PermissionHelper permissionHelper, P parameters) {
+        throw new UnsupportedOperationException("Text output not implemented");
```

Comment:
Should we return an `Optional.empty()` here and fall back to string serializing the `apply()` result instead? I think that approach would make it a bit easier to migrate tools.